torch/csrc/autograd/function.cpp

```diff
@@ -106,4 +107,65 @@ void Function::set_up_context_edge(
     backward_eval->forward_ctx_select = ctx_select;
 }
 
+/*
+ * Fix for #5534: prevent stack overflow on deletion of deep computation graph
+ *
+ * Sometimes one can end up with a very big computation graph of Functions
+ * and Edges. Each std::shared_ptr<Function> contains a list of Edge, and
+ * each Edge contains a std::shared_ptr<Function>. Deleting a
+ * std::shared_ptr<Function> can trigger the recursive deletion of other
+ * std::shared_ptr<Function>'s: this can stack overflow if the graph
+ * is deep enough. Here is an example of such a graph:
+ *
+ * shared_ptr<Function> -> Edge -> shared_ptr<Function> -> Edge -> ... -> shared_ptr<Function>
+ *
+ * The solution here is to use a custom deleter with each
+ * std::shared_ptr<Function>. The custom deleter keeps track of how many
+ * nested deleters it is in. When this number exceeds the maximum allowed
+ * depth, the Function* to be deleted are accumulated in a per-thread
+ * delete queue and handled by one of the deleters.
+ */
+thread_local std::deque<Function*> kDeleteFunctionQueue;
+
+constexpr size_t kDeleteFunctionMaxRecursionDepth = 10000;
+thread_local size_t kDeleteFunctionRecursionDepth = 0;
```

Comment:
It's a bit weird to use the k-convention here.

In the Google C++ style guide, https://google.github.io/styleguide/cppguide.html#Constant_Names the k-prefix is reserved for variables which are constant for the entire duration of the program. That is clearly not the case here.